airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationProcess.kt

```diff
@@ -0,0 +1,192 @@
+        // This could probably be a channel, somehow. But given the current structure,
+        // it's easier to just use the pipe stuff.
+        destinationStdinPipe =
+            // spotbugs requires explicitly specifying the charset,
+            // so we also have to specify autoFlush=false (i.e. the default behavior
+            // from PrintWriter(outputStream) ).
+            // Thanks, spotbugs.
+            PrintWriter(PipedOutputStream(destinationStdin), false, Charsets.UTF_8)
+        destination =
+            CliRunner.destination(
+                command,
+                config = config,
+                catalog = catalog,
+                inputStream = destinationStdin,
+            )
+    }
+
+    override fun run() {
+        destination.run()
+    }
+
+    override fun sendMessage(message: AirbyteMessage) {
+        destinationStdinPipe.println(Jsons.serialize(message))
+    }
+
+    override fun readMessages(): List<AirbyteMessage> = destination.results.newMessages()
+
+    override fun shutdown() {
+        destinationStdinPipe.close()
+    }
+}
+
+// Notably, not actually a Micronaut factory. We want to inject the actual
+// factory into our tests, not a pre-instantiated destination, because we want
+// to run multiple destination processes per test.
+// TODO only inject this when not running in CI, a la @Requires(notEnv = "CI_master_merge")
+@Singleton
+class NonDockerizedDestinationFactory : DestinationProcessFactory {
+    override fun createDestinationProcess(
+        command: String,
+        config: ConfigurationJsonObjectBase?,
+        catalog: ConfiguredAirbyteCatalog?
+    ): DestinationProcess {
+        return NonDockerizedDestination(command, config, catalog)
+    }
+}
+
+// TODO define a factory for this class + @Require(env = CI_master_merge)
+class DockerizedDestination(
+    command: String,
+    config: JsonNode?,
+    catalog: ConfiguredAirbyteCatalog?,
+) : DestinationProcess {
+    override fun run() {
+        TODO("launch a docker container")
+    }
+
+    override fun sendMessage(message: AirbyteMessage) {
+        // push a message to the docker process' stdin
+        TODO("Not yet implemented")
+    }
+
+    override fun readMessages(): List<AirbyteMessage> {
+        // read everything from the process' stdout
+        TODO("Not yet implemented")
+    }
+
+    override fun shutdown() {
+        // close stdin, wait until process exits
+        TODO("Not yet implemented")
+    }
+}
+
+// This is currently unused, but we'll need it for the Docker version.
+// it exists right now b/c I wrote it prior to the CliRunner retooling.
+/**
+ * There doesn't seem to be a built-in equivalent to this? Scanner and BufferedReader both have
+ * `hasNextLine` methods which block until the stream has data to read, which we don't want to do.
+ *
+ * This class simply buffers the next line in-memory until it reaches a newline or EOF.
+ */
+private class LazyInputStreamReader(private val input: InputStream) {
```

Comment:
seems odd that we explicitly don't want to block. I'll probably learn more below

Reply:
in case you didn't get back to this: I want something that supports this pattern:
```kotlin
val process = destinationFactory.run(...)
process.sendMessages(records, states, etc)
val messages = process.readMessages()
// verify that we ack states partway through and correctly wrote data
verifyMessages(messages)
dumpAndDiffRecords(...)
process.shutdown()
```

but a naive implementation of `readMessages` (`val s = Scanner(dockerProcess.stdout); while (s.hasNextLine()) { s.nextLine(); ... }`) will never return, because `hasNextLine` blocks forever until stdout closes (and stdout never closes, because we never make it to `process.shutdown()`).

(none of this matters until I actually implement the docker process factory though - the in-java thing uses CliRunner + BufferingOutputConsumer, which work totally fine. I might try and adapt those to work with docker, just leaving this here for now)

Reply:
ok, I see. That makes a lot of sense then. Thank you for explaining